iodata/formats/mwfn.py

```diff
@@ -0,0 +1,367 @@
+        # unrestricted wavefunction
+        num_mo = 2 * nindbasis
+
+    mo_numbers = np.empty(num_mo, int)
+    mo_type = np.empty(num_mo, int)
+    mo_occs = np.empty(num_mo, float)
+    mo_sym = np.empty(num_mo, str)
+    mo_energies = np.empty(num_mo, float)
+    mo_coeffs = np.empty([num_coeffs, num_mo], float)
+
+    for mo in range(num_mo):
+        mo_numbers[mo], mo_occs[mo], mo_energies[mo], mo_coeffs[:, mo], \
+            mo_type[mo], mo_sym[mo] = _load_helper_mo(lit, num_coeffs)
+
+    # TODO add density matrix and overlap
+
+    return {'title': title, 'energy': energy, 'wfntype': wfntype,
+            'nelec_a': nelec_a, 'nelec_b': nelec_b, 'charge': charge,
+            'atnums': atnums, 'atcoords': atcoords, 'atcorenums': atcorenums,
+            'nbasis': nbasis, 'nindbasis': nindbasis, 'nprims': nprim,
+            'nshells': nshell, 'nprimshells': nprimshell, 'full_virial_ratio': vt_ratio,
+            'shell_centers': shell_centers, 'shell_types': shell_types,
+            'prim_per_shell': prim_per_shell, 'exponents': exponent, 'coeffs': coeffs,
+            'mo_numbers': mo_numbers, 'mo_occs': mo_occs, 'mo_energies': mo_energies,
+            'mo_coeffs': mo_coeffs, 'mo_type': mo_type, 'mo_sym': mo_sym}
+
+
+def build_obasis(shell_map: np.ndarray, shell_types: np.ndarray,
+                 exponents: np.ndarray, prim_per_shell: np.ndarray,
+                 coeffs: np.ndarray,
+                 ) -> Tuple[MolecularBasis]:
+    """Based on the fchk modules basis building.
+
+    Parameters
+    -------------
+    shell_map:  np.ndarray (integer)
+        Index of what atom the shell is centered on. The mwfn file refers to this section
+        as `Shell centers`. Mwfn indices start at 1, this has been modified and starts
+        at 0 here. For water (O, H, H) with 6-31G, this would be an array like
+        [0, 0, 0, 0, 0, 1, 1, 2, 2]. , `O` in 6-31G has 5 shells and`H` has two shells.
+    shell_types: np.ndarray (integer)
+        Angular momentum of the shell. Indices start at 0 for 's' orbital, 1 for 'p' etc.
+        For 6-31G for a heavy atom this would be [0, 0, 1, 0, 1] corresponding
+         to [1s, 2s, 2p, 2s, 2p]
+    exponents: np.ndarray (float)
+        Gaussian function decay exponents for the primitives in the basis set.
+    prim_per_shell: np.ndarray (integer)
+        Array denoting the number of primitives per shell. If basis set is 6-31G this will be
+        [6, 3, 3, 1, 1] if the atom is a heavy atom. This corresponds to
+        [1s, 2s, 2p, 2s, 2p]. If additional atoms are present, the array is extended.
+    coeffs: np.ndarray (float)
+        Array of same length as `exponents` containing orbital expansion coefficients.
+    """
+    shells = []
+    counter = 0
+    # First loop over all shells
+    for i, n in enumerate(prim_per_shell):
+        shells.append(Shell(
+            shell_map[i],
+            [abs(shell_types[i])],
+            ['p' if shell_types[i] < 0 else 'c'],
```

Comment:
This seems to contradict the conventions for pure functions mentioned above?

Reply:
I will review. As noted in next comment, I was building this based on wfn.py and fchk.py and this may not be as robust as hoped.

Reply:
It was actually OK. I just misunderstood the code.